DOCS-6851-added new content (#18)
mulesoft/docs-service-mesh: modules/ROOT/pages/bind-api-configure-service-mesh-CRD.adoc

= Bind Your Service to an API Using the CRD
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: service mesh, microservices, binding

When you create the binding, Anypoint Service Mesh creates a three-way association of the service in Kubernetes,the API that you created in Anypoint Platform to represent the service, and the adapter that you created for the service. 

The adapter uses the unique API ID to fetch policies and apply them to your services that are running in the cluster.

> If you created the API in the previous procedure using Anypoint Platform, you use the API ID to bind the service. However, if you used the CRD to create the API, you must use the API Name to bind the service.

Reviewer: ```suggestion
If you created the API in the previous procedure using Anypoint Platform, you use the API ID to bind the service. However, if you used the CRD to create the API, you must use the API name to bind the service.
```

Author: Done! Thanks!

After the binding is created, Anypoint Service Mesh takes a few seconds to complete the binding procedure and update the status. After the binding is completed, you can view the status using the `list` command. The status values include:

> * Ready--The binding is now ready to be used. 

Reviewer: ditto colons

Author: Done! Thanks!

* Failed--The binding was not created successfully. An error occurred. 
* Binding--The binding is in the process of being associated with specific application.

> In the following example, a binding called “customers” is created to bind the “anypoint-service-mesh-adapter” adapter with the "customers" service. This binding is created in the “nto-payments” namespace, which is the same namespace where the adapter was previously created, and is associated with the API that has the ID as 1234.

Reviewer: Here and in the same text above, consider using ` ` rather than quotation marks.

Reviewer: End with colon.

Author: Done! Thanks!

[source,text,linenums]
----
NAMESPACE      NAME        ADAPTER                         API ID  SERVICE NAME    STATUS
nto-payments   customers   anypoint-service-mesh-adapter   1234    customers       Ready
----

[[bind_the_service]]
== Binding an API Created Using Anypoint Platform

Before you bind the service, ensure that you complete the task prerequisite.

*Task Prerequisite*
From the API Administration Tab of API Manager, obtain the API ID to create the binding using this method.

To bind the service with an API, perform the following steps:

> . Go to *Anypoint Platform > API administration*, and expand the API version name to display the instances.

Reviewer: ```suggestion
. Go to *Anypoint Platform > API administration* and expand the API version name to display the instances.
```

Author: Done! Thanks!

. Click an instance name and copy the API ID that is displayed. 
Ensure that you save this API ID to a text editor.
+
image::api-id.png[]
+
. To retrieve the service name from Kubernetes, with which you want to bind the API, run the command:
+
`kubectl get svc -n <namespace>`
. Verify that your microservice port is named following the requirements requested by Istio: xref:https://istio.io/docs/setup/additional-setup/requirements/[].
. From the Anypoint Service Mesh installer download location in the Command window, type the following command to bind the adapter with an application:
+
[source,text,linenums]
----

> ./service-mesh adapter binding create \

Reviewer 2: @priyaayyar @tbruno This section is about using CRD then why is it using a CLI command for binding creation?

Reviewer: Resolved?

Author: Done! Thanks!

--name=<binding name> \
--namespace=<namespace name> \
--adapter=<adapter name> \
--apiId=<api ID> \
--serviceName=<service name>
----

> Note that all objects names, such as adapter name, binding name, and service name, must all be in lower case. An error is thrown if you do not follow Kubernetes naming conventions.

Reviewer: Make these changes in the previous instance of this same text also:
```suggestion
Note that all object names, such as adapter name, binding name, and service name, must be in lowercase letters. An error is thrown if you do not follow Kubernetes naming conventions.
```

Author: Done! Thanks!

+
[%header%autowidth.spread]
|===
| Parameter Name | Description | Required/Optional | Default Value

> | name | The unique name that represents the binding of the Kubernetes Service with the API, for example, service-binding-anypoint-security”. 

Reviewer: ```suggestion
| name | The unique name that represents the binding of the Kubernetes Service with the API, for example, `service-binding-anypoint-security`. 
```

Author: Content removed! Thanks!

 | Required | N/A
| namespace | The namespace in which your service is deployed. Ensure that you use the same namespace as the adapter you provisioned in the previous procedure. | Required | N/A
| adapter | Specify the name of the adapter that you provisioned in the previous procedure.  | Required | N/A 
| apiId | Specify the API ID of the API that you want to associate with this adapter. Use the API ID you obtained as part of Review Prerequisites section of this document. | Required | N/A.
| serviceName | The name of the service that you want to bind to the API, which will be managed by Anypoint Platform. The service name is the identifier of the service that you want to manage with Anypoint Service Mesh. 
| Required | N/A.
|===
+

> The binding becomes activate within a few seconds and the API is successfully registered to the proxy.

Reviewer: ```suggestion
The binding becomes active within a few seconds and the API is successfully registered to the proxy.
```
OR
```suggestion
The binding is activated within a few seconds and the API is successfully registered to the proxy.
```

Author: Done! Thanks!

. Verify the binding status using the following command:
+

> `./service-mesh binding list`

Reviewer 3: ```suggestion
`./service-mesh adapter binding list`
```

Author: Done! Thanks!

+
After you bind your API, the API in API Manager is marked as Active. You are now ready to apply policies to protect your APIs.
+
image::api-manager-status-for-service-mesh.png[85%, 85%]

[[bind_using_CRD]]
== Binding the Service Using a CRD

To bind the service using a CRD:

> . Create your binding definition file and save it with a '.yaml' file extension:

Reviewer: ```suggestion
. Create your binding definition file and save it with a `.yaml` file extension.
```
Note that the formatting is performed with backticks (`), not apostrophes (').

Author: Done! Thanks!

> . If you created your API using Anypoint Platform, use spec.parameters.apiId:

Reviewer: ```suggestion
. If you created your API using Anypoint Platform, use `spec.parameters.apiId`:
```

Author: Done! Thanks!

+
[source,text,linenums]
----
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceBinding
metadata:
  name: <binding name> # example: orders-api-binding
  namespace: <namespace name> # example: default
  labels: 

>     service-mesh.mulesoft.com/kind: adapter-binding

Reviewer 2: @tbruno this seems wrong, you shouldn't need both api and adapter binding label

Reviewer 4: Yep, its wrong. In this case is adapter binding. Making suggestion to delete the second line.

Reviewer: Resolved?

Author: Done! Thanks!

>     service-mesh.mulesoft.com/kind: api-binding

Reviewer 4: ```suggestion
```

Reviewer 4: Read above.

Reviewer 3: This change it's ok

Author: Removed! Done! Thanks!

spec:
  instanceRef:
    name: <adapter name> # example: bookinfo-service-mesh-adapter
  parameters:
    apiId: <api ID> # example: 15722113
    service: <service name> # example: orders-service
----

> *Tomas: Is this ok?*

Reviewer: Resolved?

Author: Done! Thanks!

> . If you created your API using Anypoint Service Mesh, use spec.instanceRef.name:

Reviewer: ```suggestion
. If you created your API using Anypoint Service Mesh, use `spec.instanceRef.name`:
```

Author: Done! Thanks!

+
source,text,linenums]
----
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceBinding
metadata:
  # name will be used as resulting secret name if no secretName is specified under the spec section
  name: <binding name> # example: orders-api-binding

>   namespace: <namespace name> # example: default

Reviewer 3: ```suggestion
  namespace: <namespace name> # example: default
  labels:
    service-mesh.mulesoft.com/kind: api-binding
```

Author: Done! Thanks!
